PR from DEVELOP into MAIN

EmanuelRusuu opened the pull request.

EmanuelRusuu committed:
First step of the to do app. 

Completed the 1st step of learning vue: Write a basic ToDo app.

EmanuelRusuu committed:
Second step of the to do app.

Completed the 2nd step of learning vue: Extend the basic app and split it into components

EmanuelRusuu committed:
ITL-7 Create the header component

- created header component

EmanuelRusuu committed:
ITL-8 Add the title

- added "To do list" title

komaromiGyorgy commented on todoapp/src/components/TodoForm.vue:
```diff
@@ -0,0 +1,39 @@
+<script setup lang="ts">
+import { defineProps, defineEmits, ref } from 'vue'
+
+const props = defineProps(['todos', 'inputContent'])
```
Please add types to your emits and props definitions:
defineProps<{
  title?: string
  likes?: number
}>()

const emit = defineEmits<{
  (e: 'change', id: number): void
  (e: 'update', value: string): void
}>()

komaromiGyorgy commented on todoapp/src/components/TodoForm.vue:
```diff
@@ -0,0 +1,39 @@
+
+const props = defineProps(['todos', 'inputContent'])
+const emits = defineEmits(['addTodo'])
+const inputContentRef = ref(props.inputContent)
```
Please create a v-model instead of using the inputContent prop and the ref, you can create a writable computed for this purpose: 
define the appropriate props and events, and then make this style of writable computed:

const inputValue = computed({
  get() {
    return props.modelValue
  },
  set(newValue) {
    emit('update:modelValue', newValue)
  }
})

komaromiGyorgy commented on todoapp/src/components/TodoList.vue:
```diff
@@ -0,0 +1,92 @@
+  todos.value.push({
+    text: inputContent
+  })
+  inputContent = ''
```
If you empty the input content in the child component you do not need to do it here as well, after you have done the writable computed

komaromiGyorgy commented on todoapp/src/components/TodoList.vue:
```diff
@@ -0,0 +1,92 @@
+import Header from './Header.vue'
+import Todos from './Todos.vue'
+
+const todos = ref<{ text: string }[]>([])
```
You can pull out this type into a types folder and import it from there, since you use it more than one time

komaromiGyorgy commented on todoapp/src/components/TodoList.vue:
```diff
@@ -0,0 +1,92 @@
+  <div class="todoapp-container">
+    <Header @deleteTasks="deleteTasks"></Header>
+    <div class="todos-container">
+      <TodoForm :todos="todos" :inputContent="inputContent" @addTodo="addTodo"></TodoForm>
```
The TodoForm does not use the todos array, you can remove it

EmanuelRusuu committed:
ITL-9 Add an "add button"

- added the "add todo" button (included the icon from the figma design)
- changes to comply to the "Vue style guide"

komaromiGyorgy commented on todoapp/src/components/TodoUl.vue:
```diff
@@ -0,0 +1,33 @@
+
+<template>
+  <ul class="todo-ul">
+    <li class="todo" v-for="todo in todos" :key="todo.text">
```
Please strive for the key to be unique, in this case the key on the v-for could be similar, I would like to ask you to also concatenate the index of the todo with the text, so we have a higher probability of the key being unique. 
This might not matter much in your current implementation, but it will matter when you work with more complex data. Since having an unique key is crucial for the rendering engine to be able to know the difference between each node in the dom, so it can update them properly when they change.

komaromiGyorgy commented on todoapp/src/components/Todos.vue:
```diff
@@ -0,0 +1,26 @@
+<script setup lang="ts">
+import TodoUl from './TodoUl.vue'
+import { defineEmits } from 'vue'
```
You do not need to import the defineEmits

EmanuelRusuu committed:
Solved git changes request

- added types to props and emits(where needed)
- used v-model instead of props and ref 
- created a writable computed
- solved the clearing of the inputContent
- did a types folder 
- removed unnecessary code
- did an unique key for every todo concatenating the index and the todo`s value(as it`s text)

EmanuelRusuu committed:
ITL-10 Create the event that triggers the todo creation functionality

- created an add button that adds tasks when the user clicks it
- solved conflicts for 4 files complying with ITL-10

EmanuelRusuu committed:
Solved git changes request

- complied with vue style guide for type-based emits
- got rid of "any" type where needed
- replicated figma design

EmanuelRusuu committed:
Solved git changes request 2

For the first request:
- I removed the "inputValue" from the Header.vue component, because it was not needed anymore for the ITL-1.

For the second request:
- I removed the parameter "inputValue.value", from Header.vue and also from TodoForm.vue, because sending it to the parents was unnecessary.

For the third request:
- I removed all "any" types I could find, replacing them with proper ones.

For the fourth request:
- You suggested removing the "id" property from every todo in todos, and so I did. I used the index that comes with the array, and changed the removeTodo function to take "index" as parameter.

Other changes made:
- better styling;
- removed bits of unused code;
- imported the font-family used in Figma design;
- changed some code to comply with the VueStyleGuide.

EmanuelRusuu committed:
Solved git changes no-request 3

- removed unnecessary prop (v-model="inputContent") sent to Header component in TodoList.vue (ln. 38)

EmanuelRusuu committed:
Merge pull request #3 from EmanuelRusuu/feature/ITL-1-implement-page-header

ITL-1 Implement Page header
